### PR TITLE
Prevent test replacement if new test is already redeemed (EXPOSUREAPP-7447)

### PR DESCRIPTION
**Pull request by d4rken**
Issue:
User scans QR-Code successfully, later receives a link for the same QR-Code and clicks it.
The app will show the warning that the current test will be replaced.
The user confirms it, the new test registration fails (because it was already consumed), but the old test was already removed.
The user is left with no test :frowning:.

This PR:
Introduces a way to register a test, but to abort registration and keep the previous test if desired, by offering a `precondition` and `postcondition` for `registerTest`.
This is implement via `SubmissionRepository` as `registerTest` and `tryReplaceTest`.

To align errorhandling a new common class `TestRegistrationStateProcessor` handles the registration and error mapping for multiple screens.

So when scanning a new test of an already registered type, and the new test registration fails because it's already redeemed, there should still be an error dialog, but the previous test should not be removed.

Testing:
* Scan the same test twice
* Scan a qrcode, then send yourself the link for that qrcode and click it
* Normal test registration for PCR and RA should still work as before.

### CODE REVIEW

**Comment by MikeMcC399** [comment]
Link to #3354 for automatic PR status update in issue.

**Comment by MikeMcC399** [comment]
There is a typo in the title
EXPOSURAPP-7447
should be
[EXPOSUREAPP-7447](https://jira-ibs.wbs.net.sap/browse/EXPOSUREAPP-7447).
Would be good to get it corrected, because it will ultimately show up in the release log of fixes applied.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/submission/qrcode/consent/SubmissionConsentFragment.kt [hunk 1]
```diff
@@ -76,56 +70,36 @@ class SubmissionConsentFragment : Fragment(R.layout.fragment_submission_consent)
...
-                        }
-                    }
+            when (state) {
+                State.Idle,
+                State.Working -> {
+                    // Handled above
+                }
+                is State.Error -> {
+                    state.getDialogBuilder(requireContext()).show()
+                    popBackStack()
```
Comment (mtwalli) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/submission/qrcode/consent/SubmissionConsentFragment.kt:90]: is it intended to show a dialog and navigate back ?

**Review by mtwalli — COMMENTED** [review]
Probably not related to the changes in here, but registering a PENDING RAT test and then turns into Negative does not have a DCC entry in the list.

**Review by mtwalli — APPROVED** [review]
The issue here is fixed , we need to check the DCC certificate for pending negative tests

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/submission/qrcode/consent/SubmissionConsentFragment.kt (hunk 1 above)
Comment (mtwalli) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/submission/qrcode/consent/SubmissionConsentFragment.kt:90]: I checked the old code and it had the same logic  🤷

File: Corona-Warn-App/src/deviceForTesters/java/de/rki/coronawarnapp/test/coronatest/ui/CoronaTestTestFragmentViewModel.kt [hunk 2]
```diff
@@ -28,17 +24,17 @@ class CoronaTestTestFragmentViewModel @AssistedInject constructor(
 ) : CWAViewModel(dispatcherProvider = dispatcherProvider) {
 
     val errorEvents = SingleLiveEvent<Throwable>()
-    val pcrtState = coronaTestRepository.latestPCRT.map {
-        PCRTState(
-            coronaTest = it
-        )
-    }.asLiveData(context = dispatcherProvider.Default)
+    val pcrtState = coronaTestRepository.coronaTests
```
Comment (BMItter) [inline Corona-Warn-App/src/deviceForTesters/java/de/rki/coronawarnapp/test/coronatest/ui/CoronaTestTestFragmentViewModel.kt:27]: ↓ call fitler directly?

**Review by BMItr — APPROVED** [review]
lgtm! 👌 
tested on oneplus 8 device

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/viewmodel/CWAViewModel.kt [hunk 3]
```diff
@@ -27,6 +28,8 @@ abstract class CWAViewModel constructor(
         Timber.tag(tag).v("Initialized")
     }
 
+    fun <T> Flow<T>.asLiveData2() = asLiveData(context = dispatcherProvider.Default)
```
Comment (BMItter) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/viewmodel/CWAViewModel.kt:31]: maybe a more descriptive name here would be nice..

**Pull request merged**